File: android/app/src/main/java/com/kurt/bgmate/data/repository/RuleJudgeRepositoryImpl.kt
package com.kurt.bgmate.data.repository

import android.content.Context
import com.kurt.bgmate.data.local.JudgeHistoryDao
import com.kurt.bgmate.data.local.JudgeHistoryEntity
import com.kurt.bgmate.data.local.toDomain
import com.kurt.bgmate.data.remote.llm.LlmClient
import com.kurt.bgmate.data.remote.llm.LlmMessage
import com.kurt.bgmate.data.remote.llm.LlmRequest
import com.kurt.bgmate.domain.model.JudgeResult
import com.kurt.bgmate.domain.repository.RuleJudgeRepository
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject

class RuleJudgeRepositoryImpl @Inject constructor(
    private val llmClient: LlmClient,
    val judgeHistoryDao: JudgeHistoryDao,
    @ApplicationContext private val context: Context
) : RuleJudgeRepository {
    private val systemPrompt: String by lazy {
        context.assets.open("rule_judge_prompt.txt").bufferedReader().use { it.readText() }
    }

    override fun judge(
        gameName: String,
        dispute: String
    ): Flow<String> {
        val userMessage = "게임 이름: $gameName\n\n[분쟁 상황]\n$dispute"

        val request = LlmRequest(
            messages = listOf(LlmMessage(role = "user", content = userMessage)),
            systemPrompt = systemPrompt,
            maxTokens = 4096
        )

        return llmClient.stream(request)
    }

    override suspend fun saveHistory(
        gameName: String,
        dispute: String,
        answer: String
    ) {
        judgeHistoryDao.insert(
            JudgeHistoryEntity(
                gameName = gameName,
                dispute = dispute,
                answer = answer
            )
        )
    }

    override fun observeHistory(): Flow<List<JudgeResult>> =
        judgeHistoryDao.observeAll().map { list -> list.map { it.toDomain() } }
}

File: flutter/lib/data/repository/rule_judge_repository_impl.dart
import 'package:bgmate_flutter/data/remote/ai/llm_client.dart';
import 'package:bgmate_flutter/data/remote/ai/llm_request.dart';
import 'package:bgmate_flutter/domain/repository/rule_judge_repository.dart';
import 'package:flutter/services.dart';

class RuleJudgeRepositoryImpl implements RuleJudgeRepository {
  final LlmClient _llmClient;
  String? _systemPrompt;

  RuleJudgeRepositoryImpl(this._llmClient) {
    _loadPrompt();
  }

  Future<void> _loadPrompt() async {
    _systemPrompt = await rootBundle.loadString(
      'assets/prompts/rule_judge_prompt.txt',
    );
  }

  @override
  Stream<String> judge(String gameName, String question) {
    final userMessage = '게임 이름: $gameName\n\n[분쟁상황]\n$question';

    return _llmClient.stream(
      LlmRequest(
        systemPrompt: _systemPrompt,
        messages: [LlmMessage(role: 'user', content: userMessage)],
      ),
    );
  }
}
